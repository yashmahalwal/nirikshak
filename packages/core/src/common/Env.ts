/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
