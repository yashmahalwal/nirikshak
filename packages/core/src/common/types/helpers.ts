import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
