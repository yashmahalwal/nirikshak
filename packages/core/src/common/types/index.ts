export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
