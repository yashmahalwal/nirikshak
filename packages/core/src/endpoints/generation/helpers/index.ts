export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
