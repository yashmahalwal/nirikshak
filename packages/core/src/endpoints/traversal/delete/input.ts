import {
    HeaderAndQueryInstance,
    generateHeaderAndQuery,
} from "../../generation/input/headerAndQuery";
import { Inputs } from "../..";
import { ResourceInstance, SchemaHelpers } from "../../..";

export interface DeleteInput {
    semantics: HeaderAndQueryInstance;
}

// Generate input from an resource
export async function generateDeleteInput(
    input: Inputs["DELETE"],
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<DeleteInput> {
    return {
        semantics: await generateHeaderAndQuery(
            input.semantics,
            resource,
            helpers
        ),
    };
}
