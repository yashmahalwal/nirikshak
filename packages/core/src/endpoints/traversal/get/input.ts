import {
    HeaderAndQueryInstance,
    generateHeaderAndQuery,
} from "../../generation/input/headerAndQuery";
import { Inputs } from "../..";
import { ResourceInstance, SchemaHelpers } from "../../..";

export interface GetInput {
    semantics: HeaderAndQueryInstance;
}

// Generate input from an resource
export async function generateGetInput(
    input: Inputs["GET"],
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<GetInput> {
    return {
        semantics: await generateHeaderAndQuery(
            input.semantics,
            resource,
            helpers
        ),
    };
}
