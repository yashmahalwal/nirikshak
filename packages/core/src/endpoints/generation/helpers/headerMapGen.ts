import { HeaderMap } from "../../types/helpers";
import { ResourceInstance } from "../../../resource/types/helper";
import { SchemaHelpers } from "../../../common/types/helpers";
import { Literal } from "../../../common/types/literals";
import { generateBaseType } from "./baseTypeGen";

export interface HeadersInstance {
    [key: string]: Exclude<Literal, null>;
}

export async function generateHeaderMap(
    input: HeaderMap,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<HeadersInstance> {
    const headers: HeadersInstance = {};
    const promiseMap: Map<string, Promise<Literal>> = new Map();

    for (const key in input) {
        promiseMap.set(key, generateBaseType(input[key], resource, helpers));
    }

    for (const key in input) {
        const resolved = await promiseMap.get(key)!;
        if (resolved === null)
            throw new Error(
                `Header: ${key} = ${input[key]} resolved to null value`
            );
        else headers[key] = resolved;
    }

    return headers;
}
