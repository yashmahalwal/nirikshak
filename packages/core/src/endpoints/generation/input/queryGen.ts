import { Literal } from "../../../common/types/literals";
import { ResourceInstance } from "../../../resource/types/helper";
import { SchemaHelpers } from "../../../common/types/helpers";
import { generateBaseType } from "../helpers/baseTypeGen";
import { Query } from "../../types/input";

export interface QueryInstance {
    [key: string]: Exclude<Literal, null> | Exclude<Literal, null>[];
}

export async function generateQuery(
    input: Query,
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<QueryInstance> {
    const promiseMap: Map<string, Promise<Literal | Literal[]>> = new Map();

    for (const key in input) {
        const entry = input[key];

        if (Array.isArray(entry)) {
            promiseMap.set(
                key,
                Promise.all(
                    entry.map((e) => generateBaseType(e, resource, helpers))
                )
            );
        } else {
            promiseMap.set(key, generateBaseType(entry, resource, helpers));
        }
    }

    const queries: QueryInstance = {};
    for (const key in input) {
        const res = await promiseMap.get(key)!;

        if (Array.isArray(res) && res.some((entry) => entry === null))
            throw new Error(
                `query: ${key}:${input[key]} resolved to an array containing a null value`
            );
        else if (res === null)
            throw new Error(`query: ${key}:${input[key]} resolved to null`);

        queries[key] = res as Exclude<Literal, null> | Exclude<Literal, null>[];
    }

    return queries;
}
