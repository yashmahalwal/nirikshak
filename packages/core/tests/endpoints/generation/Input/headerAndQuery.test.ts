import faker from "faker";
import { SchemaHelpers } from "../../../../src/common/types/helpers";
import { HeaderAndQuery } from "../../../../src/endpoints/types/input";
import {
    HeaderAndQueryInstance,
    generateHeaderAndQuery,
} from "../../../../src/endpoints/generation/input/headerAndQuery";
import { RANDOMNESS_ITERATIONS } from "../../../../src/common/Env";
import { generateHeaderMap } from "../../../../src/endpoints/generation/helpers/headerMapGen";
import { generateQuery } from "../../../../src/endpoints/generation/input/queryGen";
const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidEntries: Array<{
    input: HeaderAndQuery;
    output: () => Promise<HeaderAndQueryInstance>;
}> = [
    {
        input: {
            headers: {
                name: 15,
                age: 55,
                word: "custom:word",
                lines: "faker:lorem.lines",
                id: "resource:id",
            },
        },
        async output(): Promise<HeaderAndQueryInstance> {
            return {
                headers: await generateHeaderMap(
                    this.input.headers,
                    resource,
                    Helpers
                ),
            };
        },
    },
    {
        input: {
            query: {
                name: 15,
                age: 55,
                word: "custom:word",
                lines: "faker:lorem.lines",
                id: "resource:id",
                purples: [
                    1,
                    false,
                    true,
                    true,
                    true,
                    false,
                    "custom:word",
                    "faker:random.word",
                    "resouce:addresses[0].zipCode",
                ],
            },
        },
        async output(): Promise<HeaderAndQueryInstance> {
            return {
                query: await generateQuery(this.input.query, resource, Helpers),
            };
        },
    },
    {
        input: {
            headers: {
                name: 15,
                age: 55,
                word: "custom:word",
                lines: "faker:lorem.lines",
                id: "resource:id",
            },
            query: {
                name: 15,
                age: 55,
                word: "custom:word",
                lines: "faker:lorem.lines",
                id: "resource:id",
                purples: [
                    1,
                    false,
                    true,
                    true,
                    true,
                    false,
                    "custom:word",
                    "faker:random.word",
                    "resouce:addresses[0].zipCode",
                ],
            },
        },
        async output(): Promise<HeaderAndQueryInstance> {
            const h: HeaderAndQueryInstance = {};

            const h1 = generateHeaderMap(this.input.headers, resource, Helpers);
            const q1 = generateQuery(this.input.query, resource, Helpers);

            h.headers = await h1;
            h.query = await q1;

            return h;
        },
    },
];

const InvalidEntries: any[] = [
    {
        headers: { a: null },
    },
    {
        query: { a: null },
    },
    {
        headers: { a: null },
        query: { a: null },
    },
];

describe("Header and Query type generation", () => {
    for (let i = 1; i <= RANDOMNESS_ITERATIONS; i++)
        ValidEntries.forEach((entry, index) => {
            test(`Valid header and query entry: ${index}, iteration: ${i}`, async () => {
                faker.seed(123 * i + index);
                const got = await generateHeaderAndQuery(
                    entry.input,
                    resource,
                    Helpers
                );
                faker.seed(123 * i + index);
                const expected = await entry.output();
                expect(got).toEqual(expected);
            });
        });

    test.each(InvalidEntries)(
        `Invalid header and query entry: %#`,
        async (entry) => {
            expect.hasAssertions();
            try {
                await generateHeaderAndQuery(entry, resource, Helpers);
            } catch (e) {
                expect(e).toMatchSnapshot();
            }
        }
    );
});
