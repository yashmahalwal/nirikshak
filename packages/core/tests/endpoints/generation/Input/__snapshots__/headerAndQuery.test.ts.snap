// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header and Query type generation Invalid header and query entry: 0 1`] = `[Error: Header: a = null resolved to null value]`;

exports[`Header and Query type generation Invalid header and query entry: 1 1`] = `[Error: query: a:null resolved to null]`;

exports[`Header and Query type generation Invalid header and query entry: 2 1`] = `[Error: Header: a = null resolved to null value]`;
